### PR TITLE
31479 updated DEM xml to take latest spec changes

### LINKED ISSUE #31479: Updated DEM XML to take latest spec changes
The Device Energy Management spec was updated to add support for Opt-out and cancelling of commands.

This URL should show all of the diffs between the versions last used:
https://github.com/CHIP-Specifications/connectedhomeip-spec/compare/c6f084b1188a44909e3d4fcb71bb272c9ac41772..ef7c371442003936081fffa0726f5b67b08334aa#diff-69ea3be1200f480acd5492108766fc8061ac57b4a2fc1336b886cb7a6dc638e9

and then look at `src/app_clusters/DeviceEnergyManagement.adoc`

This issue is to update the SDK / XML to match:

Main changes are:
* ClusterRevision: `3`
* Feature split:
`ForecastAdjustment (FA)` now subdivided into `StartTimeAdjustment (STA)`, `Pauseable (PA)`, `ForecastAdjustment (FA)`, `ConstraintBasedAdjustment (CON)`
* ESAStateEnum has removed UserOptOut
* NEW: OptOutStateEnum added
* NEW: AdjustmentCauseEnum added
* NEW: ForecastUpdateReasonEnum  added
* Updated ForecastStruct Type:
   * Added ForecastUpdateReason
   * Change to Conformance (FA now STA)
* Updated SlotStruct:
   * Change to Conformance for 3 fields (was M, now PAU)
* NEW: OptOutState Attribute   (0x0007)
* Changes to conformance for commands FA changed to STA/PAU/FA/CON as needed
* NEW: CancelRequest command
* Updated: PowerAdjustRequest takes new `Cause` enum
* Updated: StartTimeAdjustRequest takes new `Cause` enum
* Updated: PauseRequest takes new `Cause` enum
* Updated: ModifyForecastRequest takes new `Cause` enum
* Updated: RequestConstraintBasedForecast takes new `Cause` enum
* Events: 
   * Conformance for Paused/Resume is `PAU`
   * Resumed Event now includes CauseEnum

### PR DESCRIPTION
**Pull request by jamesharrow**
Fixes #31479 
 - https://github.com/project-chip/connectedhomeip/issues/31479

**Description of spec changes:** 
The Device Energy Management spec was updated to add support for Opt-out and cancelling of commands.

This URL should show all of the diffs between the versions last used:
https://github.com/CHIP-Specifications/connectedhomeip-spec/compare/c6f084b1188a44909e3d4fcb71bb272c9ac41772..ef7c371442003936081fffa0726f5b67b08334aa#diff-69ea3be1200f480acd5492108766fc8061ac57b4a2fc1336b886cb7a6dc638e9

and then look at `src/app_clusters/DeviceEnergyManagement.adoc`

This issue is to update the SDK / XML to match:

Main changes are:
* ClusterRevision: `3`
* Feature split:
`ForecastAdjustment (FA)` now subdivided into `StartTimeAdjustment (STA)`, `Pauseable (PA)`, `ForecastAdjustment (FA)`, `ConstraintBasedAdjustment (CON)`
* ESAStateEnum has removed UserOptOut
* NEW: OptOutStateEnum added
* NEW: AdjustmentCauseEnum added
* NEW: ForecastUpdateReasonEnum  added
* Updated ForecastStruct Type:
   * Added ForecastUpdateReason
   * Change to Conformance (FA now STA)
* Updated SlotStruct:
   * Change to Conformance for 3 fields (was M, now PAU)
* NEW: OptOutState Attribute   (0x0007)
* Changes to conformance for commands FA changed to STA/PAU/FA/CON as needed
* NEW: CancelRequest command
* Updated: PowerAdjustRequest takes new `Cause` enum
* Updated: StartTimeAdjustRequest takes new `Cause` enum
* Updated: PauseRequest takes new `Cause` enum
* Updated: ModifyForecastRequest takes new `Cause` enum
* Updated: RequestConstraintBasedForecast takes new `Cause` enum
* Events: 
   * Conformance for Paused/Resume is `PAU`
   * Resumed Event now includes CauseEnum

**SDK Changes in this PR**

Note XML was autogen'd by Hasty's new Alchemy tool, with minor hand edits.

 - [x] .ZAP / autogen updated
 - [x] Added xml.json to remove new OptOutState attribute from ember
 - [x] Fixed compilation issues with removal of kOptOut in ESAStateEnum
 - [x] Added OptOutState attribute read 
 - [x] Added Cause into commands that need it
 - [x] Added CancelRequest command handler
 - [x] Added more validation to optout type to reject commands from Grid vs Local
 - [x] Update FeatureMap to add new feature supported

### CODE REVIEW

**Review by andy31415 — APPROVED** [review]

File: src/app/clusters/device-energy-management-server/device-energy-management-server.cpp [hunk 1]
```diff
@@ -226,19 +236,88 @@ void Instance::InvokeCommand(HandlerContext & handlerContext)
...
-void Instance::HandlePowerAdjustRequest(HandlerContext & ctx, const Commands::PowerAdjustRequest::DecodableType & commandData)
+Status Instance::CheckOptOutAllowsRequest(AdjustmentCauseEnum adjustmentCause)
 {
-    int64_t power        = commandData.power;
-    uint32_t durationSec = commandData.duration;
-    bool validArgs       = false;
-    Status status        = Status::Success;
+    OptOutStateEnum optOutState = mDelegate.GetOptOutState();
+
+    if (adjustmentCause >= AdjustmentCauseEnum::kUnknownEnumValue)
```
Comment (bzbarsky-apple) [inline src/app/clusters/device-energy-management-server/device-energy-management-server.cpp:258]:
```suggestion
    if (adjustmentCause == AdjustmentCauseEnum::kUnknownEnumValue)
```
Valid values can be larger than kUnknownEnumValue.  What data model decoding guarantees is that all unknown values are coerced to kUnknownEnumValue.

File: src/app/clusters/device-energy-management-server/device-energy-management-server.cpp [hunk 2]
```diff
@@ -226,19 +236,88 @@ void Instance::InvokeCommand(HandlerContext & handlerContext)
...
+        return Status::InvalidValue;
+    }
+
+    switch (optOutState)
+    {
+    case OptOutStateEnum::kNoOptOut: /* User has NOT opted out so allow it */
+        ChipLogProgress(Zcl, "DEM: OptOutState = kNoOptOut");
+        return Status::Success;
+
+    case OptOutStateEnum::kLocalOptOut: /* User has opted out from Local only*/
```
Comment (bzbarsky-apple) [inline src/app/clusters/device-energy-management-server/device-energy-management-server.cpp:270]: Nothing in the spec has normative requirements that clearly correspond to this method.  That needs to get fixed in the spec.  Please make sure a spec issue is filed.

To be precise, the spec says:

> the ESA SHALL validate that ... the OptOutState permits the specified AdjustmentCauseEnum

but nothing clearly defines in a normative way what adjustment causes are allowed in which OptOutState values.  The "Summary" column for enums is not normative.

File: src/app/clusters/device-energy-management-server/device-energy-management-server.cpp [hunk 3]
```diff
@@ -226,19 +236,88 @@ void Instance::InvokeCommand(HandlerContext & handlerContext)
...
+        ChipLogProgress(Zcl, "DEM: OptOutState = kLocalOptOut");
+        switch (adjustmentCause)
+        {
+        case AdjustmentCauseEnum::kGridOptimization:
+            return Status::Success;
+        case AdjustmentCauseEnum::kLocalOptimization:
+        default:
```
Comment (bzbarsky-apple) [inline src/app/clusters/device-energy-management-server/device-energy-management-server.cpp:277]: Not having a default (but having a case for kUnknownEnumValue, which you already checked above, I know) will make this not compile when a new cause value is added, which might be desirable.  Similar in other switch statements.

File: src/app/zap-templates/zcl/data-model/chip/device-energy-management-cluster.xml [hunk 4]
```diff
@@ -113,74 +127,104 @@ limitations under the License.
...
     <item name="Offline" value="0x01"/>
     <item name="Fault" value="0x02"/>
     <item name="UserOptOut" value="0x03"/>
+    <item name="Cancelled" value="0x04"/>
+  </enum>
+
+  <enum name="AdjustmentCauseEnum" type="enum8">
+    <cluster code="0x0098"/>
+    <item name="LocalOptimization" value="0x00"/>
+    <item name="GridOptimization" value="0x01"/>
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/device-energy-management-cluster.xml:152]: Note: In the spec these have spaces.  Please fix the spec.

File: src/app/zap-templates/zcl/data-model/chip/device-energy-management-cluster.xml [hunk 5]
```diff
@@ -113,74 +127,104 @@ limitations under the License.
...
+  </enum>
+
+  <enum name="ForecastUpdateReasonEnum" type="enum8">
+    <cluster code="0x0098"/>
+    <item name="InternalOptimization" value="0x00"/>
+    <item name="LocalOptimization" value="0x01"/>
+    <item name="GridOptimization" value="0x02"/>
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/device-energy-management-cluster.xml:159]: Again, in the spec these have spaces.  Please fix the spec.

File: src/app/clusters/device-energy-management-server/device-energy-management-server.cpp (hunk 3 above)
Comment (jamesharrow) [reply src/app/clusters/device-energy-management-server/device-energy-management-server.cpp:277]: Ok will amend

File: src/app/zap-templates/zcl/data-model/chip/device-energy-management-cluster.xml (hunk 4 above)
Comment (jamesharrow) [reply src/app/zap-templates/zcl/data-model/chip/device-energy-management-cluster.xml:152]: https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/8766 raised

File: src/app/zap-templates/zcl/data-model/chip/device-energy-management-cluster.xml (hunk 5 above)
Comment (jamesharrow) [reply src/app/zap-templates/zcl/data-model/chip/device-energy-management-cluster.xml:159]: https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/8766 raised

File: src/app/clusters/device-energy-management-server/device-energy-management-server.cpp (hunk 2 above)
Comment (jamesharrow) [reply src/app/clusters/device-energy-management-server/device-energy-management-server.cpp:270]: https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/8766 raised

**Review by bzbarsky-apple — APPROVED** [review]

**Comment by michaelleecowan** [comment]
Approve, not on official XML review lst so a 'grey' check

**Pull request merged**